README.md
# composer-module-setup

A plugin for Composer to declare automatically jelix modules into a jelix application

For Jelix 1.6.9 and higher.

Author who provide their modules via Composer, should declare directories 
containing modules or plugins. It will avoid the developer to declare them 
into his application.init.php (Jelix 1.7) or in the configuration (Jelix 1.6.9+).

## installation

In the composer.json of your application, declare the plugin

```json
{
    "require": {
        "jelix/composer-module-setup": "^1.0.0"
    }
}
```

## Declaring modules and plugins into a package

Author who provide their modules via Composer, should declare directories containing modules
or plugins. It will avoid the developer to declare them into his application.init.php.

To declare them, he should add informations into the extra/jelix object in composer.json:

```json
{
    "extra": {
        "jelix": {
        }
    }
}
```

In this object, he can add three type of informations:

- `modules-dir`, an array containing directories where modules can be found.
  These paths will be added to the jelix configuration parameter `modulesPath`.
- `plugins-dir`, an array containing directories where plugins can be found.
  These paths will be added to the jelix configuration parameter `pluginsPath`.
- `modules`, an array containing modules directories.
  These paths will be added into the `modules` section of the jelix configuration
   as `<module>.path=<path>`.

For instances, in the repository, if modules are in a "modules/" directory, the 
author should add these informations into his composer.json:

```json
{
    "extra": {
        "jelix": {
            "modules-dir" : [
                "modules/"
            ],
            "plugins-dir" : [
                "plugins/"
            ],
            "modules" : [
                "mymoduleaaa/",
                "mymodulebbb/"
            ]
        }
    }
}
```

## Declaring modules and plugins at the application level

Application developers could declare also their modules and plugins in the same way, in
the composer.json of the application:

```json
{
    "require": {
        "jelix/composer-module-setup": "^1.0.0"
    },
    "extra": {
        "jelix": {
            "modules-dir" : [
                "myapp/modules/"
            ],
            "plugins-dir" : [
                "myapp/plugins/"
            ],
            "modules" : [
                 "mainmodule/"
            ]
        }
    }
}
```


## Enable modules on entrypoints automatically (jelix 1.6 only)

In module packages, you can indicate on which entrypoint the module should be enabled.

```json
{
    "extra": {
        "jelix": {
            "autoconfig-access-16" : {
                "__any_app" : {
                    "<modulename>": {
                        "__global": 1
                    }
                },
                "app/identifiant1" : {
                    "<modulename>": {
                        "__global": 1,
                        "index" : 2,
                        "admin" : 1   
                    }
                }
            }
        }
    }
}
```

`app/identifiant1` must be the application id that is indicated into the `project.xml`
file ( `id` attribute of the `<info>` element).

In the composer.json of the application, you can also indicate the same informations
for each modules, when a module does not provide an `"autoconfig-access-16"` configuration 

```json
{
    "extra": {
        "jelix": {
            "modules-autoconfig-access-16" : {
                "package/name1" : {
                    "<modulename>": {
                        "__global": 1,
                        "index" : 2,
                        "admin" : 1   
                    }
                },
                "package/name2" : {
                    "<modulename>": {
                        "__global": 1,
                        "index" : 2,
                        "admin" : 1   
                    }
                }
            }
        }
    }
}
```

If the package has already an `"autoconfig-access-16"` configuration, this
`modules-autoconfig-access-16` configuration has priority over it.

## In Jelix 1.7 and higher

In your application.init.php, you should include the jelix_app_path.php:

```php
<?php

require (__DIR__.'/vendor/autoload.php');

jApp::initPaths(
    __DIR__.'/'
    //__DIR__.'/www/',
    //__DIR__.'/var/',
    //__DIR__.'/var/log/',
    //__DIR__.'/var/config/',
    //__DIR__.'/scripts/'
);
jApp::setTempBasePath(realpath(__DIR__.'/temp').'/');

require (__DIR__.'/vendor/jelix_app_path.php');

```

Remember: in Jelix 1.7 and higher, declaring modules and plugins in the modulesPath/pluginsPath
parameter in the configuration file is not supported anymore.

## In Jelix 1.6.x equal or higher than 1.6.9

The composer plugin declares automatically modules and plugins directory into 
the localconfig.ini.php file or into the mainconfig.ini.php file, 
in `modulesPath` and `pluginsPath` properties, and also in the `modules` section.

However, at the application level, the composer.json may also content the path
to the application directory (the directory containing the project.xml etc), and
the path to the var/config directory.


So if the directory containing the composer.json file is not the application 
directory and/or if the var/config is not in the application directory, you must 
set these paths into `app-dir` and `var-config-dir`. Path should be
relative to the composer.json directory, or can be absolute.

```json
{
    "require": {
        "jelix/composer-module-setup": "^1.0.0"
    },
    "extra": {
        "jelix": {
            "app-dir" : "myapp/",
            "var-config-dir" : "/var/lib/myapp/config/",
            "modules-dir" : []
        }
    }
}
```

You can also indicate if you want that the plugin modify localconfig.ini.php 
or mainconfig.ini.php. By default it is localconfig.ini.php. Indicate the
configuration filename into `config-file-16`:


```json
{
    "require": {
        "jelix/composer-module-setup": "^1.0.0"
    },
    "extra": {
        "jelix": {
            "app-dir" : "myapp/",
            "var-config-dir" : "/var/lib/myapp/config/",
            "modules-dir" : [],
            "config-file-16": "mainconfig.ini.php"
        }
    }
}
```

## debugging the plugin

Set an environnement variable `JELIX_DEBUG_COMPOSER` to `true` or create an 
empty file named `JELIX_DEBUG_COMPOSER` into the vendor directory.

After launching Composer, you will have a file `jelix_debug.log` into
the vendor directory.


CHANGELOG.md
Version 1.0.12
==============

Fix Jelix 1.7 setup when no app exists yet.

Version 1.0.11
===============

Fix Jelix 1.7 configurator: the configuration parameters `<module>.path` set
by the Jelix 1.6 configurator should be removed.

Version 1.0.10
==============

Fix an error when updating the packages, whereas there is no application yet.

Version 1.0.9
=============

Fix previous version: it did not use the new internal api `isJelix16()` :-/

Version 1.0.8
=============

Possibility to force the target jelix version for migration

When migrating an app from Jelix 1.6 to Jelix 1.7, the application is still
with the files of Jelix 1.6. So the setup is not done with Jelix 1.7.
Specifying the target allow to launch the setup for the right Jelix version.

Version 1.0.7
=============

Fix: do not use SetupJelix16, when the plugin is used in a secondary composer.json

If the composer.json is not the main application composer.json, the plugin should verify on which Jelix version is
based the application, with a better method than checking jelix packages, as into
a secondary composer.json don't load the jelix packages.


Version 1.0.6
=============

- Fix: do not check the file declared into `config-file-16`, as it may not exists for Jelix 1.7

Version 1.0.5
=============

- Fix path into jelix_modules_infos.json, when there are some file links
- Improve message logs

Version 1.0.4
=============

- New feature: possibility to have a log file of what did the plugin.
  To generate logs, set the environment variable `JELIX_DEBUG_COMPOSER=true`
- Fix access setup on an undefined default entrypoint
- Fix: when removing a module, some infos still remained into jelix app config

Version 1.0.3
=============

- Fix Jelix 1.6 configuration: activation of a module on an entrypoint should
  be removed if a new version of the module does not use anymore the entry-point

Version 1.0.2
=============

- Fix Jelix 1.6 configuration: url engine configuration was missing

Version 1.0.1
=============

- Fix: Jelix 1.6 configuration should be updated when a package is removed

Version 1.0.0
=============

- Compatibility with Composer 2.0


Version 0.7.0
=============

- Support of automatic setup of module access in the configuration for Jelix 1.6.
  See `autoconfig-access-16` and `modules-autoconfig-access-16` parameters.


Version 0.6.1
=============

- Fix installation of modules when the package name contains upper case letters


lib/JelixParameters.php
<?php

namespace Jelix\ComposerPlugin;
use Composer\Util\Filesystem;
use Composer\Package\PackageInterface;

class JelixParameters {

    const VERSION = 1;

    /**
     * @var JelixPackageParameters[]
     */
    protected $packagesInfos = array();

    /**
     * @var JelixPackageParameters[]
     */
    protected $removedPackagesInfos = array();

    protected $fs;

    /**
     * @var string
     */
    protected $vendorDir;

    /**
     * @var string
     */
    protected $varConfigDir;

    /**
     * @var string
     */
    protected $appDir = null;

    protected $autoconfig16 = array();

    protected $jelixTarget = '';

    /**
     * name of the configuration file that will be modified to declare
     * module paths. Only for Jelix 1.6
     * @var string $configurationFileName
     */
    protected $configurationFileName = '';

    function __construct($vendorDir)
    {
        $this->fs = new Filesystem();
        $this->vendorDir = rtrim($vendorDir, '/').'/';
    }

    function loadFromFile($filepath)
    {
        $content = json_decode(file_get_contents($filepath), true);

        if (!isset($content['packages'])) {
            return;
        }

        if (isset($content['target-jelix-version'])) {
            $this->jelixTarget = $content['target-jelix-version'];
        }

        foreach($content['packages'] as $package => $infos) {
            $content = array_merge_recursive(
                array(
                    'is-app'=>false,
                    'modules-dirs'=>array(),
                    'plugins-dirs'=>array(),
                    'modules'=>array(),
                    'autoconfig-access-16'=>array(),
                    'modules-autoconfig-access-16'=>array()
                ),
                $infos
            );
            $parameters = new JelixPackageParameters($package, $content['is-app']);
            $parameters->setModulesDirs($content['modules-dirs']);
            $parameters->setPluginsDirs($content['plugins-dirs']);
            $parameters->setSingleModuleDirs($content['modules']);
            $parameters->setAppModulesAccess($content['autoconfig-access-16']);
            $parameters->setPackageModulesAccess($content['modules-autoconfig-access-16']);
            $this->packagesInfos[$package] = $parameters;
        }
    }

    function saveToFile($filepath)
    {
        $content = array(
            'version'=>1,
            'target-jelix-version' => $this->jelixTarget,
            'packages'=>array(),
            ''
        );
        foreach($this->packagesInfos as $package => $parameters) {
            $content['packages'][$package] = array(
                'modules-dirs'=>$parameters->getModulesDirs(),
                'plugins-dirs'=>$parameters->getPluginsDirs(),
                'modules'=>$parameters->getSingleModuleDirs(),
                'autoconfig-access-16'=>$parameters->getAppModulesAccess(),
                'modules-autoconfig-access-16'=>$parameters->getPackageModulesAccess(),
            );
        }
        file_put_contents($filepath, json_encode($content, JSON_PRETTY_PRINT));
    }

    function getAppDir() {
        return $this->appDir;
    }

    function getVarConfigDir() {
        return $this->varConfigDir;
    }

    function getConfigFileName() {
        return $this->configurationFileName;
    }

    function getVendorDir() {
        return $this->vendorDir;
    }

    /**
     * @param string $packageName package composer name
     * @param array $extra content of the extra section of the composer.json of the package
     * @param string $packagePath path to the package
     *
     */
    function addApplicationPackage($packageName, $extra, $packagePath)
    {
        $parameters = new JelixPackageParameters($packageName, true);
        $this->packagesInfos[$packageName] = $parameters;

        if (isset($this->removedPackagesInfos[$packageName])) {
            unset($this->removedPackagesInfos[$packageName]);
        }

        if (!isset($extra['jelix'])) {
            $this->appDir = $packagePath;
            $this->appDir = rtrim($this->appDir, "/") . "/";
            $this->varConfigDir = $this->appDir . 'var/config/';

            return;
        }

        // This is the package of the application.
        // Process information from the composer.json of the application

        if (isset($extra['jelix']['app-dir'])) {
            if ($this->fs->isAbsolutePath($extra['jelix']['app-dir'])) {
                $this->appDir = $extra['jelix']['app-dir'];
            } else {
                $this->appDir = $this->fs->normalizePath($packagePath . DIRECTORY_SEPARATOR . $extra['jelix']['app-dir']);
            }
            if (!$this->appDir || !file_exists($this->appDir)) {
                throw new ReaderException("Error in composer.json of " . $packageName . ": extra/jelix/app-dir is not set or does not contain a valid path");
            }
            if (!file_exists($this->appDir . '/project.xml')) {
                throw new ReaderException("Error in composer.json of " . $packageName . ": extra/jelix/app-dir is not a path to a Jelix application");
            }

        } else {
            $this->appDir = $packagePath;
        }
        $this->appDir = rtrim($this->appDir, "/") . "/";

        $this->varConfigDir = $this->appDir . 'var/config/';

        if (isset($extra['jelix']['var-config-dir'])) {
            if ($this->fs->isAbsolutePath($extra['jelix']['var-config-dir'])) {
                $this->varConfigDir = $extra['jelix']['var-config-dir'];
            } else {
                $this->varConfigDir = $this->fs->normalizePath($packagePath . DIRECTORY_SEPARATOR . $extra['jelix']['var-config-dir']);
            }
            if (!$this->varConfigDir || !file_exists($this->varConfigDir)) {
                throw new ReaderException("Error in composer.json of " . $packageName . ": extra/jelix/var-config-dir is not set or does not contain a valid path");
            }
            $this->varConfigDir = rtrim($this->varConfigDir, "/") . "/";
        }

        if (isset($extra['jelix']['target-jelix-version']) && preg_match("/^(\\d+\\.\\d+)/", $extra['jelix']['target-jelix-version'], $v)) {
            $this->jelixTarget = $v[1];
        } else if (file_exists($this->appDir . 'app/system/mainconfig.ini.php')) {
            $this->jelixTarget = '1.7';
        } else if (file_exists($this->varConfigDir . 'mainconfig.ini.php')) {
            $this->jelixTarget = '1.6';
        }

        if (isset($extra['jelix']['config-file-16'])) {
            $this->configurationFileName = $extra['jelix']['config-file-16'];
        }

        if (isset($extra['jelix']['modules-autoconfig-access-16'])) {
            $modulesAccess = $extra['jelix']['modules-autoconfig-access-16'];
            if (is_array($modulesAccess)) {
                $cleanedModulesAccess = array();
                foreach ($extra['jelix']['modules-autoconfig-access-16'] as $package => $moduleAccess) {
                    if (!is_string($package) || !is_array($moduleAccess)) {
                        continue;
                    }
                    foreach ($moduleAccess as $module => $access) {
                        if (is_array($access)) {
                            if (!array_key_exists($package, $cleanedModulesAccess)) {
                                $cleanedModulesAccess[$package] = array();
                            }
                            $cleanedModulesAccess[$package][$module] = $access;
                        }
                    }
                }
                $parameters->setPackageModulesAccess($cleanedModulesAccess);
            }
        }

        // read information that can be in any composer.json
        $this->readModuleDirs($packageName, $packagePath, $parameters, $extra);
    }


    /**
     * @param string $packageName package composer name
     * @param array $extra content of the extra section of the composer.json of the package
     * @param string $packagePath path to the package
     */
    function addPackage($packageName, $extra, $packagePath)
    {
        $parameters = new JelixPackageParameters($packageName, false);
        $this->packagesInfos[$packageName] = $parameters;

        if (isset($this->removedPackagesInfos[$packageName])) {
            unset($this->removedPackagesInfos[$packageName]);
        }

        if (!isset($extra['jelix'])) {
            return;
        }

        // read information from the composer.json of a module package
        $this->readAutoconfigAccess($parameters, $extra);

        // read information that can be in any composer.json
        $this->readModuleDirs($packageName, $packagePath, $parameters, $extra);
    }


    protected function readModuleDirs($packageName, $packagePath, JelixPackageParameters $parameters, &$extra)
    {

        if (isset($extra['jelix']['modules-dir'])) {
            if (!is_array($extra['jelix']['modules-dir'])) {
                throw new ReaderException("Error in composer.json of ".$packageName.": extra/jelix/modules-dir is not an array");
            }
            foreach($extra['jelix']['modules-dir'] as $path) {
                $path = $packagePath.DIRECTORY_SEPARATOR.$path;
                $parameters->addModulesDir($this->fs->findShortestPath($this->vendorDir, $path, true));
            }
        }

        if (isset($extra['jelix']['modules'])) {
            if (!is_array($extra['jelix']['modules'])) {
                throw new ReaderException("Error in composer.json of " . $packageName . ": extra/jelix/modules is not an array");
            }
            foreach($extra['jelix']['modules'] as $path) {
                $path = $packagePath.DIRECTORY_SEPARATOR.$path;
                $parameters->addSingleModuleDir($this->fs->findShortestPath($this->vendorDir, $path, true));
            }
        }

        if (isset($extra['jelix']['plugins-dir'])) {
            if (!is_array($extra['jelix']['plugins-dir'])) {
                throw new ReaderException("Error in composer.json of ".$packageName.": extra/jelix/plugins-dir is not an array");
            }
            foreach($extra['jelix']['plugins-dir'] as $path) {
                $path = $packagePath.DIRECTORY_SEPARATOR.$path;
                $parameters->addPluginsDir($this->fs->findShortestPath($this->vendorDir, $path, true));
            }
        }
    }

    /**
     * read informations from the composer.json of a module package
     *
     * @param JelixPackageParameters $parameters
     * @param $extra
     * @return void
     */
    protected function readAutoconfigAccess(JelixPackageParameters $parameters, $extra)
    {
        if (isset($extra['jelix']['autoconfig-access-16'])) {
            $modulesAccess = $extra['jelix']['autoconfig-access-16'];
            if (is_array($modulesAccess)) {
                $cleanedModulesAccess = array();
                foreach($extra['jelix']['autoconfig-access-16'] as $app => $moduleAccess) {
                    if (!is_string($app) || !is_array($moduleAccess)) {
                        continue;
                    }
                    foreach($moduleAccess as $module =>$access) {
                        if (is_array($access)) {
                            if (!array_key_exists($app, $cleanedModulesAccess)) {
                                $cleanedModulesAccess[$app] = array();
                            }
                            $cleanedModulesAccess[$app][$module] = $access;
                        }
                    }
                }
                $parameters->setAppModulesAccess($cleanedModulesAccess);
            }
        }
    }

    function removePackage($packageName, $extra) {
        if(isset($this->packagesInfos[$packageName]))
        {
            unset($this->packagesInfos[$packageName]);
        }

        $parameters = new JelixPackageParameters($packageName, false);
        $this->removedPackagesInfos[$packageName] = $parameters;
        $this->readAutoconfigAccess($parameters, $extra);
    }

    function getPackageParameters($packageName) {
        if(isset($this->packagesInfos[$packageName]))
        {
            return $this->packagesInfos[$packageName];
        }
        return null;
    }

    /**
     * @return JelixPackageParameters[]
     */
    function getPackages()
    {
        return $this->packagesInfos;
    }

    /**
     * @return JelixPackageParameters[]
     */
    function getRemovedPackages()
    {
        return $this->removedPackagesInfos;
    }

    /**
     * @return JelixPackageParameters
     */
    function getApplicationPackage()
    {
        foreach($this->packagesInfos as $package)
        {
            if ($package->isApp()) {
                return $package;
            }
        }
        return null;
    }

    function getAllModulesDirs() {
        $allModulesDir = array();
        foreach( $this->packagesInfos as $packageName => $parameters) {
            $allModulesDir = array_merge($allModulesDir, $parameters->getModulesDirs());
        }
        return $allModulesDir;
    }

    function getAllPluginsDirs() {
        $allPluginsDir = array();
        foreach( $this->packagesInfos as $packageName => $parameters) {
            $allPluginsDir = array_merge($allPluginsDir, $parameters->getPluginsDirs());
        }
        return $allPluginsDir;
    }

    function getAllSingleModuleDirs() {
        $allModules = array();
        foreach( $this->packagesInfos as $packageName => $parameters) {
            $allModules = array_merge($allModules, $parameters->getSingleModuleDirs());
        }
        return $allModules;
    }

    function isJelix16()
    {
        if ($this->jelixTarget == '') {
            if ($this->getPackageParameters('jelix/jelix') ||
                $this->getPackageParameters('jelix/jelix-essential') ||
                $this->getPackageParameters('jelix/for-classic-package')  // deprecated
            ) {
                return false;
            }

            if (file_exists($this->appDir.'app/system/mainconfig.ini.php')) {
                $this->jelixTarget = '1.7';
                return false;
            }

            if (file_exists($this->varConfigDir . 'mainconfig.ini.php')) {
                $this->jelixTarget = '1.6';
                return true;
            }
        }
        else if ($this->jelixTarget == '1.6') {
            return true;
        }
        return false;
    }
}
